rosidl_generator_py/resource/_import_type.c.em

```diff
@@ -0,0 +1,32 @@
+@{
+module_name, type_name = full_type_name.rsplit(".", maxsplit=1)
+ensure_is_type = locals().get('ensure_is_type', False)
+}@
+@
+static PyObject * @(func_name)()
+{
+  // TODO(almaslov) proper deinitialization
```

Comment:
Sorry for late response.
I’ve researched a little about best practice for this case, and in short, I think it will reqire changing public API.
First let me explain what is this comment about. Deserialization function should import a message class to be able to construct its instance. In current upstream solution it is imported on every call. So if large message contains 1000 submessages, than import machinery is called 1000 times too. And thought repeated calls of module import is relatively cheap, in total 1000 imports are quite expensive. `_import_type.c.em` is a naive implementation of cache which stores cached value in static C variable. First problem here is that imported type objects are not decref’ed if support module unimported and will be garbage collected only on interpreter exit. The second problem potential issues with python subinterpreters ([some info here](https://peps.python.org/pep-0554/)).

I think these imported type objects should be stored in some state-object to be finalized later. It could be implemented in some ways
* Python3 has recommende way to store global state - PyModule_GetState, which is described here: https://peps.python.org/pep-3121/. 
* rclpy bindings could pass some special state-object and manage its lifetime to implement more complicated logic. 

As I can see, additional parameter with state-object required by `convert_to_py` and `convert_from_py` C methods in any way.
What do you think about it?

Reply:
Sorry for the late response as well,  I now understand a little better the use of `_import_type.c.em` and the issue at hand. The general idea of the module seems fine to me. In regards to the issue at hand,  after reading a bit about the solution you propose I like the  idea in order to avoid memory leaks. The only thing that worries me is the modification of the public API, Do you have a general idea of how much changes this would imply?

Another possible solution could be to manually handle the decref of the imported type objects by using `Py_DECREF()` at the moment of unloading the module. Do you think it could work? Let me know what you think.

Reply:
@almaslov and I discussed this a bit offline.  I can't think of any instance where we are using sub-interpreters, so this should _likely_ work.  I think that we should add a comment in the code to indicate that this is unsupported in the case of sub-interpreters.  

I think that the likely scenario is that you would end up with a double-free in that case (though that is a guess), and we should save someone some time in the future if they run into this unsupported use case.